File: Scala/data_structures/src/main/scala-2.12/binary_search_trees/ValidBinarySearchTree.scala
package binary_search_trees

object ValidBinarySearchTree {
  import scala.annotation.tailrec

  type Stack[T] = List[T]

  final case class Node(key: Int, leftIx: Int, rightIx: Int)

  class BinaryTree(nodes: Vector[Node]) {
    private val rootIndex: Int = 0

    def inorderTraversal(): List[Int] = {
      @tailrec
      def loop(keys: List[Int], nodeIndex: Int, stack: Stack[Node]): List[Int] = {
        if (nodeIndex != -1) {
          val node: Node = nodes(nodeIndex)
          loop(keys, node.leftIx, node :: stack)
        }
        else stack match {
          case Nil => keys.reverse
          case Node(key, _, rightIx) :: rest => loop(key :: keys, rightIx, rest)
        }
      }
      loop(Nil, rootIndex, Nil)
    }
  }

  private def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  private[binary_search_trees] def readNodeInformation(reader: Iterator[String], nrNodes: Int): Vector[Node] = {
    (for { _ <- 0 until nrNodes } yield {
      val List(key, left, right): List[Int] = convertToIntList(reader.next())
      Node(key, left, right)
    }).toVector
  }

  private def isSorted(lst: List[Int], length: Int): Boolean =
    length <= 1 || lst.sliding(2).forall { case List(a, b) => a <= b }

  def isValidBinarySearchTree(tree: BinaryTree, nrNodes: Int): Boolean = {
    if (nrNodes <= 1) true
    else {
      val inorderKeys: List[Int] = tree.inorderTraversal()
      isSorted(inorderKeys, nrNodes)
    }
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val nrNodes: Int = reader.next().toInt
    val nodes: Vector[Node] = readNodeInformation(reader, nrNodes)
    val tree = new BinaryTree(nodes)
    val verdict: Boolean = isValidBinarySearchTree(tree, nrNodes)
    println(if (verdict) "CORRECT" else "INCORRECT")
  }
}

File: OCaml/data_structures/binary_search_trees/valid_binary_search_tree.ml

type node = { key : int; left_ix : int; right_ix : int }

module type BinaryTreeSig =
    sig
        type t
        val create : node array -> int -> t
        val inorder_traversal : t -> int list
    end

module BinaryTree : BinaryTreeSig =
    struct
        type t = { nodes : node array; root_ix : int }

        let create nodes root_ix = { nodes; root_ix }

        let inorder_traversal { nodes; root_ix } =
            let rec loop keys node_index stack =
                if node_index <> -1 then
                    let node = nodes.(node_index) in
                    loop keys node.left_ix (node :: stack)
                else match stack with
                    | [] -> List.rev keys
                    | { key; right_ix } :: rest -> loop (key :: keys) right_ix rest
            in loop [] root_ix []
    end


let convert_to_intlist line = List.map int_of_string Str.(line |> split (regexp " "))


let read_node_information nr_nodes =
    let rec loop nodes k =
        if k = nr_nodes then Array.of_list @@ List.rev nodes
        else
            let key, left, right = Scanf.sscanf (read_line()) "%d %d %d" (fun x y z -> x, y, z) in
            loop ({ key; left_ix = left; right_ix = right} :: nodes) (k + 1)
    in loop [] 0


let rec is_sorted = function
    | [] | [_] -> true
    | a :: (b :: rest as lst) -> if a <= b then is_sorted lst else false


let is_valid_binary_search_tree tree nr_nodes =
    if nr_nodes <= 1 then true
    else
        let inorder_keys = BinaryTree.inorder_traversal tree in
        is_sorted inorder_keys


let () =
    let nr_nodes = read_int() in
    let nodes = read_node_information nr_nodes in
    let tree = BinaryTree.create nodes 0 in
    let verdict = is_valid_binary_search_tree tree nr_nodes in
    print_endline @@ if verdict then "CORRECT" else "INCORRECT"
